example/onvif-util.1
.\" Text automatically generated by txt2man
.TH ONVIF-UTIL  "1" "25 October 2022" ""
.SH NAME 

onvif-util
.SH SYNOPSIS
.nf
.fam C

  \fBonvif-util\fP [\fB-ah\fP] [\fB-u\fP <user>] [\fB-p\fP <password>] [\fIhost_ip_address\fP]

.fam T
.fi
.fam T
.fi
.SH DESCRIPTION

View and set parameters on onvif compatible IP cameras. The command may be used to find and identify cameras, and then to create an interactive session that can be used to query and set camera properties. 
.PP
\fB-a\fP, \fB--all\fP
show all cameras on the network
.PP
\fB-h\fP, \fB--help\fP
show the help for this command
.PP
\fB-u\fP, \fB--user\fP 
set the username for the camera login
.PP
\fB-p\fP, \fB--password\fP
set the password for the camera login
.PP
To view all cameras on the network:
\fBonvif-util\fP \fB-a\fP
.PP
To login to a particular camera:
\fBonvif-util\fP \fB-u\fP username \fB-p\fP password ip_address
.PP
To login to a camera with safe mode disabled:
\fBonvif-util\fP \fB-s\fP \fB-u\fP username \fB-p\fP password ip_address
.PP
Once logged into the camera you can view data using the 'get' command followed by the data requested. The (n) indicates an optional profile index to apply the setting, otherwise the current profile is used
.PP
.nf
.fam C
        Data Retrieval Commands (start with get)

        get rtsp 'pass'(optional) (n) - Get rtsp uri for camera, with optional password credential
        get capabilities
        get time
        get profiles
        get profile (n)
        get video (n)
        get video options (n)
        get imaging
        get imaging options
        get network

        Parameter Setting Commands (start with set)

        set resolution (n) - Resolution setting in the format widthxheight, must match option
        set framerate (n)
        set gov_length (n)
        set bitrate (n)
        set bightness value(required)
        set contrast value(required)
        set saturation value(required)
        set sharpness value(required)
        set ip_address value(required)
        set default_gateway value(required)
        set dns value(required)
        set dhcp value(required) - Accepted settings are 'on' and off'
        set password  value(required)

        Maintenance Commands

        help
        safe - set safe mode on.  Viewer and browser are disabled
        unsafe - set safe mode off.  Viewer and browser are enabled
        browser - Use browser to access camera configurations
        view (n) - View the camera output using ffplay (ffplay must be installed in the path)
        view player (n) - View the camera output with user specified player e.g. view vlc
        sync_time 'zone'(optional) - Sync the camera time to the computer
        dump - Full set of raw data from camera configuration
        reboot

        To Exit Camera Session

        quit

.fam T
.fi
.SH EXAMPLES

A typical session would begin by finding the cameras on the network
.PP
> \fBonvif-util\fP \fB-a\fP
.PP
.nf
.fam C
      Looking for cameras on the network\.\.\.
      Found 8 cameras
      192.168.1.18 localhost(TV TV-IP319PI)
      192.168.1.7 (IPC-BO IPC-122)
      192.168.1.14 IPC(Dahua IPC-HDW4631C-A)
      192.168.1.6 IPC(Amcrest IP2M-841EB)
      192.168.1.12 (AXIS M1065-LW)
      192.168.1.12 (AXIS M1065-LW)
      192.168.1.2 IPC(Amcrest IP3M-HX2W)
      192.168.1.11 R2(IPC-model)

.fam T
.fi
To start a session with a camera, use the login credentials
.PP
> \fBonvif-util\fP \fB-u\fP admin \fB-p\fP admin123 192.168.1.12
.PP
.nf
.fam C
      found host: 192.168.1.12
      successfully connected to host
        name:   AXIS M1065-LW
        serial: ACCC8E99C915

.fam T
.fi
Get current settings for video
.PP
> get video
.PP
.nf
.fam C
      Profile set to profile_1_h264

      Resolution: 1920 x 1080
      Frame Rate: 25
      Gov Length: 30
      Bit Rate:   4096

.fam T
.fi
Get available video settings
.PP
> get video options
.PP
.nf
.fam C
      Available Resolutions
        1920 x 1080
        1280 x 720
        640 x 480
        320 x 240
      Min Gov Length: 1
      Max Gov Length: 32767
      Min Frame Rate: 1
      Max Frame Rate: 30
      Min Bit Rate: 1
      Max Bit Rate: 2147483647

.fam T
.fi
Set video resolution
.PP
> set resolution 1280x720
.PP
.nf
.fam C
      Resolution was set to 1280 x 720

.fam T
.fi
Exit session
.PP
> quit

.SH COPYRIGHT
Copyright \(co 2022 Stephen Rhodes
License GPLv2+: GNU GPL version 2 or later <https://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO 

There is a GUI version of this program included with the libonvif package which will implement most of the same commands. It may be invoked using the 'onvif-camadmin' command. The GUI has the ability to view camera video output using a player such as ffplay, provided that the player executable is installed in the computer path.
.SH NOTES

Camera compliance with the onvif standard is often incomplete and in some cases incorrect. Success with the \fBonvif-util\fP may be limited in many cases. Cameras made by Hikvision will have the greatest level of compatibility with \fBonvif-util\fP. Cameras made by Dahua will have a close degree of compatability with some notable exceptions regarding gateway and DNS settings. Time settings may not be reliable in some cases. If the time is set without the zone flag, the time appearing in the camera feed will be synced to the computer time. If the time zone flag is used, the displayed time may be set to an offset from the computer time based on the timezone setting of the camera.
.PP
If the camera DNS setting is properly onvif compliant, the IP address may be reliably set using \fBonvif-util\fP. Some cameras may not respond to the DNS setting requested by \fBonvif-util\fP due to non compliance. Note that the camera may reboot automatically under some conditions if the DNS setting is changed from off to on.
.PP
Video settings are reliable. The Admin Password setting is reliable, as well as sync_time and the reboot command. If there is an issue with a particular setting, it is recommended to connect to the camera using the browser command, as most cameras will have a web interface that will allow you to make the changes reliably. The GUI version has a button on the Admin tab that will launch the web browser with the camera ip address automatically.


debian/changelog
libonvif (1.3.0-1) unstable; urgency=medium

  * Initial release (Closes: #1021980).

 -- Petter Reinholdtsen <pere@debian.org>  Tue, 18 Oct 2022 10:35:17 +0200


example/onvif-util.c
/*******************************************************************************
* onvif-util.c
*
* Copyright (c) 2022 Stephen Rhodes 
* Originally authored by Brian D Scott
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*
*******************************************************************************/

#include <stdio.h>
#include <string.h>
#include "onvif.h"
#ifdef _WIN32
#include "getopt-win.h"
#else
#include <getopt.h>
#endif

#define DEFAULT_USERNAME    "admin"
#define DEFAULT_PASSWORD    "admin"
int longopt = 0;

static struct option longopts[] = {
             { "user",      required_argument,	NULL,	'u'},
             { "password",      required_argument,	NULL,	'p'},
             { "all",      no_argument,	NULL,	'a'},
             { "rate",      required_argument,	NULL,	'r'},
             { "width",      required_argument,	NULL,	'w'},
             { "height",      required_argument,	NULL,	'h'},
#ifdef ONVIF1906
             { "guaranteed",      no_argument,	NULL,	'g'},
             { "noguaranteed",      no_argument,	NULL,	'G'},
             { "no-guaranteed",      no_argument,	NULL,	'G'},
#endif
             { "name",      required_argument,	&longopt,	1},
             { "turbo",      no_argument,	&longopt,	2},
             { "normal",      no_argument,	&longopt,	3},
             { NULL,         0,                      NULL,           0 }
     };

static const char *username = DEFAULT_USERNAME;
static const char *password = DEFAULT_PASSWORD;
static int all = 0;
static int rate = -1;
static int width = -1;
static int height = -1;
#ifdef ONVIF1906
static int guaranteed = -1;
#endif
static int setName = 0;
static char *name;
static int turbo = -1;
/*
	Multicast settings?
*/

static void processOne(struct OnvifData *onvif_data);

int main(int argc, char **argv)
{
	int ch;
	struct OnvifSession *onvif_session = (struct OnvifSession*)malloc(sizeof(struct OnvifSession));
    struct OnvifData *onvif_data = (struct OnvifData*)malloc(sizeof(struct OnvifData));
	char *arg0 = argv[0];

	while ((ch = getopt_long(argc, argv, "u:p:ar:w:h:gG", longopts, NULL)) != -1) {
		switch (ch) {
            case 'u':
				username = optarg;
				break;
			case 'p':
				password = optarg;
				break;
			case 'a':
				all = 1;
				break;
			case 'r':
				rate = atoi(optarg);
				break;
			case 'w':
				width = atoi(optarg);
				break;
			case 'h':
				height = atoi(optarg);
				break;
#ifdef ONVIF1906
			case 'g':
				guaranteed = 1;
				break;
			case 'G':
				guaranteed = 0;
				break;
#endif
			case 0:
				if (longopt == 1) {
					setName = 1;
					name = optarg;
				} else if (longopt == 2) {
					turbo = 1;
				} else if (longopt == 3) {
					turbo = 0;
				}
				longopt = 0;
				break;
			default:
				exit(1);
		}
	}
	argc -= optind;
    argv += optind;

    if (!all && argc < 1) {
        fprintf(stderr, "Usage: %s <options> [HOST ..]\n", arg0);
        exit(1);
    } else if (all && argc > 0) {
        fprintf(stderr, "Usage: %s <options> [HOST ..]\n", arg0);
        exit(1);
	}
    initializeSession(onvif_session);
    int n = broadcast(onvif_session);

	if (all && n == 0) {
		fprintf(stderr,"No cameras on this network\n");
		exit(1);
	}

	if (all) {
		for (int i = 0; i < n; i++) {
			prepareOnvifData(i, onvif_session, onvif_data);
			processOne(onvif_data);
		}
	} else {
		while (argc--) {
			char *wanted = argv++[0];
			int found = 0;
			for (int i = 0; i < n; i++) {
				prepareOnvifData(i, onvif_session, onvif_data);
				char host[128];
				extractHost(onvif_data->xaddrs, host);
				if (strcmp(host,wanted) == 0) {
					found = 1;
					break;
				}
			}
			if (found) {
				processOne(onvif_data);
			} else {
				fprintf(stderr,"Camera at %s not found\n",wanted);
			}
		}
	}
}

static void processOne(struct OnvifData *onvif_data) {
	char host[128];
	extractHost(onvif_data->xaddrs, host);
	getHostname(onvif_data);
	printf("%s %s(%s)\n",host,
		onvif_data->host_name,
		onvif_data->camera_name);
	if (
#ifdef ONVIF1906
		guaranteed >= 0 ||
#endif
		turbo >= 0 || rate > 0 || width > 0 || height > 0) {
		getFirstProfileToken(onvif_data);
		getProfile(onvif_data);
		getVideoEncoderConfiguration(onvif_data);
#ifdef ONVIF1906
		if (guaranteed == 1)
			onvif_data->guaranteed_frame_rate = true;
		else if (guaranteed == 0)
			onvif_data->guaranteed_frame_rate = false;
#endif
		if (rate > 0)
			onvif_data->frame_rate = rate;
		if (width > 0)
			onvif_data->width = width;
		if (height > 0)
			onvif_data->height = height;

		if (turbo > 0) {
/*
	BitrateLimit (kbps)
	EncodingInterval (1)
	Quality (1-6 on mine: Should ideally check VideoEncoderConfigurationOptions for real range.)
*/
			onvif_data->conf_bitrate_limit = onvif_data->bitrate_max;
			onvif_data->conf_encoding_interval = 1;
			onvif_data->quality = 6.0;
		} else if (turbo == 0) {
			onvif_data->conf_bitrate_limit = 4096;
			onvif_data->conf_encoding_interval = 1;
			onvif_data->quality = 4.0;
		}

		if (setVideoEncoderConfiguration(onvif_data) < 0) {
			fprintf(stderr,"Error: %s\n",onvif_data->last_error);
		}
	}
	if (setName) {
		strcpy(onvif_data->host_name,name);
		if (setHostname(onvif_data) < 0) {
			fprintf(stderr,"Error: %s\n",onvif_data->last_error);
		}
	}
}